make cross-platform compilation (build.sh) work for rama-cli

`build.sh` was added in #410

It does however not work out of the box on macos.

Would be great it if can be fixed such that we can easily build for all these platforms and get the binaries for upload into a (github) release.
If needed it can be done via docker, but I believe with tooling like _zig_ it should be possible to do so without the need for docker.

The difficulty for now however is Cloudflare's boring-sys...

https://github.com/plabayo/rama/blob/main/rama-cli/scripts/build.sh

@GlenDC Hi, I am looking into this issue, as you said, I tried to run the build command on Apple M1:

```
os=macos rama-cli/scripts/build.sh
```

and got this error:
```
/Users/airycanon/Projects/rust/rama/target/aarch64-apple-darwin/debug/build/boring-sys-c827d412ac814934/out/boringssl/src/crypto/cpu_aarch64_apple.c:56:2: error: "NEON and crypto extensions should be statically available."
     56 | #error "NEON and crypto extensions should be statically available."
        |  ^
  1 error generated.
  make[3]: *** [CMakeFiles/crypto.dir/src/crypto/cpu_aarch64_apple.c.o] Error 1
  make[3]: *** Waiting for unfinished jobs....
  make[2]: *** [CMakeFiles/crypto.dir/all] Error 2
  make[1]: *** [CMakeFiles/ssl.dir/rule] Error 2
  make: *** [ssl] Error 2

  thread 'main' panicked at /Users/airycanon/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/cmake-0.1.54/src/lib.rs:1119:5:

  command did not execute successfully, got: exit status: 2

  build script failed, must exit now
```

After some debugging, I found that the compilation flags for boring differ between build methods.

When compiling boring-sys directly (which completes successfully), these flags are used:

```
CMAKE_BUILD_PARALLEL_LEVEL="10"  
CMAKE_PREFIX_PATH=""  
LC_ALL="C"  
"cmake"  
"/Users/airycanon/Projects/rust/boring/target/debug/build/boring-sys-9e35a02476267827/out/boringssl"  
"-DCMAKE_OSX_ARCHITECTURES=arm64"  
"-DCMAKE_INSTALL_PREFIX=/Users/airycanon/Projects/rust/boring/target/debug/build/boring-sys-9e35a02476267827/out"  
"-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=15.2"  
"-DCMAKE_C_COMPILER=/usr/bin/cc" "-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=15.2"  
"-DCMAKE_CXX_COMPILER=/usr/bin/c++" "-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=15.2"  
"-DCMAKE_ASM_COMPILER=/usr/bin/cc" "-DCMAKE_BUILD_TYPE=Debug"
```

However, when compiling boring-sys with zigbuild in rama, different flags are used:

```
CMAKE_BUILD_PARALLEL_LEVEL="10"  
CMAKE_PREFIX_PATH=""  
LC_ALL="C"  
"cmake"  
"/Users/airycanon/Projects/rust/rama/target/aarch64-apple-darwin/debug/build/boring-sys-c827d412ac814934/out/boringssl"  
"-DCMAKE_OSX_ARCHITECTURES=arm64"  
"-DCMAKE_TOOLCHAIN_FILE=/Users/airycanon/Library/Caches/cargo-zigbuild/0.19.8/cmake/aarch64-apple-darwin-toolchain.cmake"  
"-DCMAKE_INSTALL_PREFIX=/Users/airycanon/Projects/rust/rama/target/aarch64-apple-darwin/debug/build/boring-sys-c827d412ac814934/out"  
"-DCMAKE_C_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=15.2"  
"-DCMAKE_CXX_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=15.2"  
"-DCMAKE_ASM_FLAGS= -ffunction-sections -fdata-sections -fPIC --target=arm64-apple-macosx -mmacosx-version-min=15.2"  
"-DCMAKE_BUILD_TYPE=Release"  
```

The zigbuild method uses a custom toolchain `zigbuild/0.19.8/cmake/aarch64-apple-darwin-toolchain.cmake` with the following content:

```
set(CMAKE_SYSTEM_NAME Darwin)
set(CMAKE_SYSTEM_PROCESSOR arm64)
set(CMAKE_C_COMPILER /Users/airycanon/Library/Caches/cargo-zigbuild/0.19.8/zigcc-aarch64-apple-darwin-d40.sh)
set(CMAKE_CXX_COMPILER /Users/airycanon/Library/Caches/cargo-zigbuild/0.19.8/zigcxx-aarch64-apple-darwin-d40.sh)
set(CMAKE_RANLIB /Users/airycanon/Library/Caches/cargo-zigbuild/0.19.8/zigranlib.sh)
set(CMAKE_AR /Users/airycanon/Library/Caches/cargo-zigbuild/0.19.8/ar)
```

The compiler scripts are as follows:

```
CMAKE_C_COMPILER
#!/bin/sh
exec "/Users/airycanon/.cargo/bin/cargo-zigbuild" zig cc -- -g -target aarch64-macos-none "$@"

CMAKE_CXX_COMPILER
#!/bin/sh
exec "/Users/airycanon/.cargo/bin/cargo-zigbuild" zig c++ -- -g -target aarch64-macos-none "$@"
```

Both methods use the `cargo-zigbuild zig` command to call the actual compiler.

Then I looked into the zigbuild source code to understand how it handles compilation flags. I noticed that the `cargo-zigbuild` implementation specifically handles the `--march` option in its code: [https://github.com/rust-cross/cargo-zigbuild/blob/main/src/zig.rs#L327](https://github.com/rust-cross/cargo-zigbuild/blob/main/src/zig.rs#L327)

Based on this understanding, I tried setting explicit CFLAGS and CXXFLAGS with the correct architecture and crypto extensions. Using the following command successfully resolved the build error:

```
os=macos CFLAGS="-march=apple_m1+crypto" CXXFLAGS="-march=apple_m1+crypto" rama-cli/scripts/build.sh
```

The error is now resolved, but a remaining challenge is how to properly determine the correct `-march` value for different systems. Since this solution works specifically for Apple M1 with `apple_m1+crypto`, we would need a systematic approach to identify the appropriate architecture flags for other platforms.

The error is now resolved, but a remaining challenge is how to properly determine the correct `-march` value for different systems. Since this solution works specifically for Apple M1 with `apple_m1+crypto`, we would need a approach to identify the appropriate architecture flags for other macos platforms.

Do you have any suggestion?

--- a/.github/workflows/release.yaml
+++ b/.github/workflows/release.yaml
@@ -1,0 +1,162 @@
+name: Build Releases
+on:
+  push:
+    tags:
+      - v*
+  workflow_dispatch:
+    inputs:
+      tag:
+        description: 'Release Tag'
+        required: true
+        type: string
+
+env:
+  CARGO_TERM_COLOR: always
+  RUST_TOOLCHAIN: stable
+
+jobs:
+  build-release-linux:
+    runs-on: ubuntu-latest
+    env:
+      RUST_BACKTRACE: full
+    strategy:
+      fail-fast: true
+      matrix:
+        target:
+          - name: x86_64-unknown-linux-gnu
+            features: mimalloc
+          - name: aarch64-unknown-linux-gnu
+            features: mimalloc
+          - name: armv7-unknown-linux-gnueabi
+            features: jemalloc
+          - name: armv7-unknown-linux-gnueabihf
+            features: jemalloc
+          - name: arm-unknown-linux-gnueabi
+            features: jemalloc
+          - name: arm-unknown-linux-gnueabihf
+            features: jemalloc
+          - name: i686-unknown-linux-gnu
+            features: jemalloc
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{env.RUST_TOOLCHAIN}}
+          targets: ${{ matrix.target.name }}
+
+      - uses: Swatinem/rust-cache@v2
+        with:
+          env-vars: "RUST_TOOLCHAIN=${{env.RUST_TOOLCHAIN}}"
+
+      - name: Install cross
+        run: |
+          cargo install --git https://github.com/cross-rs/cross --branch main cross
+
+      - name: Build release
+        timeout-minutes: 120
+        run: |
+          cross build --release -p rama-cli --target ${{ matrix.target.name }} --features ${{ matrix.target.features }}
+
+      - name: Package
+        run: |
+          cd target/${{ matrix.target.name }}/release && tar -cJf rama.${{ matrix.target.name }}.tar.xz rama
+
+      - name: Upload Github Assets
+        uses: softprops/action-gh-release@v2
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          files: target/${{ matrix.target.name }}/release/*.tar.xz
+          prerelease: ${{ contains(github.ref_name, '-') }}
+          tag_name: ${{ inputs.tag || github.ref_name }}
+
+  build-release-macos:
+    runs-on: macos-latest
+    env:
+      RUST_BACKTRACE: full
+    strategy:
+      fail-fast: true
+      matrix:
+        target:
+          - x86_64-apple-darwin
+          - aarch64-apple-darwin
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Install GNU tar
+        if: runner.os == 'macOS'
+        run: |
+          brew install gnu-tar
+          echo "/usr/local/opt/gnu-tar/libexec/gnubin" >> $GITHUB_PATH
+
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{env.RUST_TOOLCHAIN}}
+          targets: ${{ matrix.target }}
+
+      - uses: Swatinem/rust-cache@v2
+        with:
+          env-vars: "RUST_TOOLCHAIN=${{env.RUST_TOOLCHAIN}}"
+
+      - name: Build release
+        shell: bash
+        run: |
+          cargo build --release -p rama-cli --target ${{ matrix.target }}
+
+      - name: Package
+        run: |
+          cd target/${{ matrix.target }}/release && tar -cJf rama.${{ matrix.target }}.tar.xz rama
+
+      - name: Upload Github Assets
+        uses: softprops/action-gh-release@v2
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          files: target/${{ matrix.target }}/release/*.tar.xz
+          prerelease: ${{ contains(github.ref_name, '-') }}
+          tag_name: ${{ inputs.tag || github.ref_name }}
+
+  build-release-windows:
+    runs-on: ubuntu-latest
+    env:
+      RUST_BACKTRACE: full
+    strategy:
+      fail-fast: true
+      matrix:
+        target:
+          - x86_64-pc-windows-gnu
+          - i686-pc-windows-gnu
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: dtolnay/rust-toolchain@master
+        with:
+          toolchain: ${{env.RUST_TOOLCHAIN}}
+          targets: ${{ matrix.target }}
+
+      - uses: Swatinem/rust-cache@v2
+        with:
+          env-vars: "RUST_TOOLCHAIN=${{env.RUST_TOOLCHAIN}}"
+
+      - name: Install cross
+        run: |
+          cargo install --git https://github.com/cross-rs/cross --branch main cross
+
+      - name: Build release
+        run: |
+          cross build --release -p rama-cli --target ${{ matrix.target }}
+
+      - name: Package
+        run: |
+          cd target/${{ matrix.target }}/release && tar -cJf rama.${{ matrix.target }}.tar.xz rama.exe
+
+      - name: Upload Github Assets
+        uses: softprops/action-gh-release@v2
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          files: target/${{ matrix.target }}/release/*.tar.xz
+          prerelease: ${{ contains(github.ref_name, '-') }}
+          tag_name: ${{ inputs.tag || github.ref_name }}

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -95,6 +95,7 @@ rustls = { version = "0.23", default-features = false, features = [
     "tls12",
     "aws_lc_rs",
 ] }
+aws-lc-sys = { version = "0.27", features = ["bindgen"] }
 derive_more = "2.0.1"
 rustls-native-certs = "0.8.0"
 rustls-pemfile = "2.1"

--- a/Cross.toml
+++ b/Cross.toml
@@ -1,0 +1,114 @@
+[target.armv7-unknown-linux-gnueabi]
+pre-build = [
+    "apt update && apt install -y gcc-arm-linux-gnueabi g++-arm-linux-gnueabi"
+]
+
+[target.armv7-unknown-linux-gnueabi.env]
+passthrough = [
+    "CC=arm-linux-gnueabi-gcc",
+    "CXX=arm-linux-gnueabi-g++",
+    "CMAKE_C_COMPILER=arm-linux-gnueabi-gcc",
+    "CMAKE_CXX_COMPILER=arm-linux-gnueabi-g++",
+    "CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_LINKER=arm-linux-gnueabi-gcc"
+]
+
+[target.armv7-unknown-linux-gnueabihf]
+pre-build = [
+    "apt update && apt install -y gcc-arm-linux-gnueabihf g++-arm-linux-gnueabihf"
+]
+
+[target.armv7-unknown-linux-gnueabihf.env]
+passthrough = [
+    "CC=arm-linux-gnueabihf-gcc",
+    "CXX=arm-linux-gnueabihf-g++",
+    "CMAKE_C_COMPILER=arm-linux-gnueabihf-gcc",
+    "CMAKE_CXX_COMPILER=arm-linux-gnueabihf-g++",
+    "CARGO_TARGET_ARMV7_UNKNOWN_LINUX_GNUEABIHF_LINKER=arm-linux-gnueabihf-gcc"
+]
+
+[target.arm-unknown-linux-gnueabi]
+pre-build = [
+    "apt update && apt install -y gcc-arm-linux-gnueabi g++-arm-linux-gnueabi"
+]
+
+[target.arm-unknown-linux-gnueabi.env]
+passthrough = [
+    "CC=arm-linux-gnueabi-gcc",
+    "CXX=arm-linux-gnueabi-g++",
+    "CMAKE_C_COMPILER=arm-linux-gnueabi-gcc",
+    "CMAKE_CXX_COMPILER=arm-linux-gnueabi-g++",
+    "CARGO_TARGET_ARM_UNKNOWN_LINUX_GNUEABIHF_LINKER=arm-linux-gnueabi-gcc"
+]
+
+[target.arm-unknown-linux-gnueabihf]
+pre-build = [
+    "apt update && apt install -y gcc-arm-linux-gnueabihf g++-arm-linux-gnueabihf"
+]
+
+[target.arm-unknown-linux-gnueabihf.env]
+passthrough = [
+    "CC=arm-linux-gnueabihf-gcc",
+    "CXX=arm-linux-gnueabihf-g++",
+    "CMAKE_C_COMPILER=arm-linux-gnueabihf-gcc",
+    "CMAKE_CXX_COMPILER=arm-linux-gnueabihf-g++",
+    "CARGO_TARGET_ARM_UNKNOWN_LINUX_GNUEABIHF_LINKER=arm-linux-gnueabihf-gcc"
+]
+
+[target.aarch64-unknown-linux-gnu]
+pre-build = [
+    "apt update && apt install -y gcc-aarch64-linux-gnu g++-aarch64-linux-gnu"
+]
+
+[target.aarch64-unknown-linux-gnu.env]
+passthrough = [
+    "CC=aarch64-linux-gnu-gcc",
+    "CXX=aarch64-linux-gnu-g++",
+    "CMAKE_C_COMPILER=aarch64-linux-gnu-gcc",
+    "CMAKE_CXX_COMPILER=aarch64-linux-gnu-g++",
+    "CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_LINKER=aarch64-linux-gnu-gcc"
+]
+
+[target.i686-unknown-linux-gnu]
+pre-build = [
+    "apt update && apt install -y gcc-i686-linux-gnu g++-i686-linux-gnu"
+]
+
+[target.i686-unknown-linux-gnu.env]
+passthrough = [
+    "CC=i686-linux-gnu-gcc",
+    "CXX=i686-linux-gnu-g++",
+    "CMAKE_C_COMPILER=i686-linux-gnu-gcc",
+    "CMAKE_CXX_COMPILER=i686-linux-gnu-g++",
+    "CARGO_TARGET_I686_UNKNOWN_LINUX_GNU_LINKER=i686-linux-gnu-gcc"
+]
+
+[target.x86_64-pc-windows-gnu]
+pre-build = [
+    "apt update && apt install -y libclang-10-dev clang-10 cmake nasm gcc-mingw-w64-x86-64 g++-mingw-w64-x86-64"
+]
+
+[target.x86_64-pc-windows-gnu.env]
+passthrough = [
+    "CC=x86_64-w64-mingw32-gcc",
+    "CXX=x86_64-w64-mingw32-g++",
+    "CMAKE_C_COMPILER=x86_64-w64-mingw32-gcc",
+    "CMAKE_CXX_COMPILER=x86_64-w64-mingw32-g++",
+    "CARGO_TARGET_X86_64_PC_WINDOWS_GNU_LINKER=x86_64-w64-mingw32-gcc",
+    "BINDGEN_EXTRA_CLANG_ARGS=-idirafter/usr/include"
+]
+
+[target.i686-pc-windows-gnu]
+pre-build = [
+    "apt update && apt install -y libclang-10-dev clang-10 cmake nasm gcc-mingw-w64-i686 g++-mingw-w64-i686"
+]
+
+[target.i686-pc-windows-gnu.env]
+passthrough = [
+    "CC=i686-w64-mingw32-gcc",
+    "CXX=i686-w64-mingw32-g++",
+    "CMAKE_C_COMPILER=i686-w64-mingw32-gcc",
+    "CMAKE_CXX_COMPILER=i686-w64-mingw32-g++",
+    "CARGO_TARGET_I686_PC_WINDOWS_GNU_LINKER=i686-w64-mingw32-gcc",
+    "BINDGEN_EXTRA_CLANG_ARGS=-idirafter/usr/include"
+]
+

--- a/rama-net/Cargo.toml
+++ b/rama-net/Cargo.toml
@@ -17,11 +17,12 @@ workspace = true
 default = []
 http = ["dep:rama-http-types", "dep:sha2", "dep:itertools", "dep:hex"]
 tls = ["dep:hex", "dep:md5", "dep:sha2", "dep:itertools"]
-rustls = ["tls", "dep:rustls"]
+rustls = ["tls", "dep:rustls", "dep:aws-lc-sys"]
 boring = ["tls", "dep:boring", "dep:nom"]
 telemetry = ["rama-core/telemetry"]
 
 [dependencies]
+aws-lc-sys = { workspace = true, optional = true, features = ["bindgen"] }
 base64 = { workspace = true }
 boring = { workspace = true, optional = true }
 bytes = { workspace = true }

--- a/rama-net/Cargo.toml
+++ b/rama-net/Cargo.toml
@@ -22,7 +22,7 @@ boring = ["tls", "dep:boring", "dep:nom"]
 telemetry = ["rama-core/telemetry"]
 
 [dependencies]
-aws-lc-sys = { workspace = true, optional = true, features = ["bindgen"] }
+aws-lc-sys = { workspace = true, optional = true }
 base64 = { workspace = true }
 boring = { workspace = true, optional = true }
 bytes = { workspace = true }

